docs/book/how-to/overview/step-output-typing-and-annotation.md

```diff
@@ -0,0 +1,109 @@
+
+{% hint style="warning" %}
+ZenML provides a built-in [CloudpickleMaterializer](https://sdkdocs.zenml.io/latest/core\_code\_docs/core-materializers/#zenml.materializers.cloudpickle\_materializer.CloudpickleMaterializer) that can handle any object by saving it with [cloudpickle](https://github.com/cloudpipe/cloudpickle). However, this is not production-ready because the resulting artifacts cannot be loaded when running with a different Python version. In such cases, you should consider building a [custom Materializer](../handle-data-artifacts/handle-custom-data-types.md#custom-materializers) to save your objects in a more robust and efficient format.
+
+Moreover, using the `CloudpickleMaterializer` could allow users to upload of any kind of object. This could be exploited to upload a malicious file, which could execute arbitrary code on the vulnerable system.
+{% endhint %}
+
+```python
+from typing import Tuple
+from zenml import step
+
+@step
+def square_root(number: int) -> float:
+    return number ** 0.5
+
+# To define a step with multiple outputs, use a `Tuple` type annotation
+@step
+def divide(a: int, b: int) -> Tuple[int, int]:
+    return a // b, a % b
+```
+
+{% hint style="info" %}
+It is impossible for ZenML to detect whether you want your step to have a single output artifact of type `Tuple` or multiple output artifacts just by looking at the type annotation.
+
+We use the following convention to differentiate between the two: When the `return` statement is followed by a tuple literal (e.g. `return 1, 2` or `return (value_1, value_2)`) we treat it as a step with multiple outputs. All other cases are treated as a step with a single output of type `Tuple`.
+
+```python
+from zenml import step
+from typing_extensions import Annotated
+from typing import Tuple
+
+# Single output artifact
+@step
+def my_step() -> Tuple[int, int]:
+    output_value = (0, 1)
+    return output_value
+
+# Single output artifact with variable length
+@step
+def my_step(condition) -> Tuple[int, ...]:
+    if condition:
+        output_value = (0, 1)
+    else:
+        output_value = (0, 1, 2)
+
+    return output_value
+
+# Single output artifact using the `Annotated` annotation
+@step
+def my_step() -> Annotated[Tuple[int, ...], "my_output"]:
+    return 0, 1
+
+
+# Multiple output artifacts
+@step
+def my_step() -> Tuple[int, int]:
+    return 0, 1
+
+
+# Not allowed: Variable length tuple annotation when using
+# multiple output artifacts
+@step
+def my_step() -> Tuple[int, ...]:
+    return 0, 1
+```
+{% endhint %}
+
```

Comment:
maybe an image of the DAG vsiualizer here showing artifact names and types might make sense?